### ISSUE TITLE
Provide an access to the name property

### ISSUE THREAD

**Issue by monolithed** [marcohamersma/react-bem-helper#8]
### The problem:

```ts
let block = new ReactBlock({
	name: 'foo'
});

block.name // undefined

let { className, name } = block({ modifiers: ['bar'] });

name // undefined
className // .foo .foo--bar
```

#### Expected 

```ts
let block = new ReactBlock({
	name: 'foo'
});

block.name // foo

let { name } = block({ modifiers: ['bar'] });

name // foo
```

#### Before

```ts
import * as ReactBlock from 'react-bem-helper';
import * as ReactCSSTransitionGroup from 'react-addons-css-transition-group';

class Preloader extends React.Component<void, void> {
	name = 'foo';

	block = new ReactBlock({ name: this.name })

	render (): JSX.Element {
		let block = this.block({ modifiers: ['bar'] });

		return <div { ...block } >
			<ReactCSSTransitionGroup transitionName={ this.name } >
				Loading...
			</ReactCSSTransitionGroup>
		</div>;
	}
}
```

#### After

```ts
import * as ReactBlock from 'react-bem-helper';
import * as ReactCSSTransitionGroup from 'react-addons-css-transition-group';

class Preloader extends React.Component<void, void> {
	block = new ReactBlock({ name: 'foo' })

	render (): JSX.Element {
		return <div { ...this.block({ modifiers: ['bar'] } } >
			<ReactCSSTransitionGroup transitionName={ this.block.name } >
				Loading...
			</ReactCSSTransitionGroup>
		</div>;
	}
}
```

**Comment by marcohamersma** [comment]
This makes sense, I'll see if I can pick this up in a future version.

**Comment by AlexanderOtavka** [comment]
With #7, you will be able to do this:

```jsx
import * as ReactBlock from 'react-bem-helper';
import * as ReactCSSTransitionGroup from 'react-addons-css-transition-group';

class Preloader extends React.Component<void, void> {
	block = new ReactBlock({ name: 'foo', outputIsString: true })

	render (): JSX.Element {
		return <div className={ this.block({ modifiers: ['bar'] } } >
			<ReactCSSTransitionGroup transitionName={ this.block() } >
				Loading...
			</ReactCSSTransitionGroup>
		</div>;
	}
}
```

Do you think this solves your issue?

**Comment by monolithed** [comment]
> Do you think this solves your issue?

I like your proposal, but it looks like over configuration in the current release. Such changes should be worked by default.  

```js
block = new ReactBlock({ name: 'foo', outputIsString: true });
```

vs

```js
block = new ReactBlock('foo');
```

In the other hand my changes can be released without breaking backward compatibility.

**Comment by AlexanderOtavka** [comment]
Indeed.  Our solution is to export a new function `withDefaults`.  It creates a new constructor with custom defaults set.  To export these defaults to your whole project, create a new module like so:
```js
// ReactBlock.js
import {withDefaults} from 'react-bem-helper';
export default withDefaults({
  outputIsString: true,
  // what any other default values for the constructor
});
```

Then, we can do your solution like so; no backwards compatibility broken.

```js
import ReactBlock from './ReactBlock';

const block = new ReactBlock('foo');
```

And it will work as if you passed `outputIsString` into the constructor.

**Comment by monolithed** [comment]
> To export these defaults to your whole project, create a new module like 

I see what you mean, but you solution is not user friendly for deep directory structures. 
I my case, it would look like:

```js
import ReactBlock from '../../../../utils/ReactBlock';
```

 It's pain for future code refactoring..

**Comment by AlexanderOtavka** [comment]
Hmm.  Would a `BEMHelper.setGlobalDefaults` function be better?  Maybe I could add that for project-wide defaults, and people can use `withDefaults` for more scoped defaults, such as a default prefix for only some parts of the package.

The only reason I am resistant to your approach is that I want the `outputIsString` anyway (for my own project), and I'd rather kill two birds with one stone.

**Comment by monolithed** [comment]
> Would a BEMHelper.setGlobalDefaults function be better? 

Maybe.

> The only reason I am resistant to your approach is that I want the outputIsString anyway (for my own project), and I'd rather kill two birds with one stone.

I guess our proposals are very different and I don't see any sense to continue this discussion. Because what if I want to get the other params such as `extra` , `modifiers` and `elements`? It costs nothing and can be useful for many people.

**Comment by AlexanderOtavka** [comment]
Alright, that's reasonable.  In my PR, I'm already exposing a `block.modifiers` function that can be used like so:

```jsx
const block = new ReactBlock('foo');

block.modifiers('bar', 'baz'); // -> { className: 'foo__bar--baz' }
```

I could tweak it a little to always return a string (regardless of the value of `outputIsString`) and add the complementary `block.element()` function like so:

```jsx
block.modifiers('bar', 'baz'); // -> 'foo__bar--baz'
block.element('bar'); // -> 'foo__bar'
block.element(''); // -> 'foo'
```

I'll do that in a separate PR though.  In the meantime, I'll add `setGlobalDefaults()` to my current one in case people with deeply nested code bases like yours want to change other project defaults.

**Comment by AlexanderOtavka** [comment]
Actually, I won't add `setGlobalDefaults` because it is brittle and impossible to typecheck.  Maybe webpack aliases are the way to go for deeply nested project structures.

Also, I'll add `block.element` and `block.modifiers` in my PR now.

**Comment by marcohamersma** [comment]
Sorry for my lack of chiming in about this, I'm having some visitors at the moment and have no free time available. I hope to look at it either sunday or monday.

**Comment by marcohamersma** [comment]
I think I'll add make the options accessible through the helper's instance, like so:

```js
const helper = new BEMHelper('componentName');

console.log(helper.name) // " componentName;"
console.log(helper.modifierDelimiter) //  '--'
```

This will be done either as a part of #7 or afterwards.

@monolithed I don't really see a good use case for exposing things like `extra` , `modifiers` and `elements` right now. What did you have in mind?

That said, implying that “It costs nothing“ is a bit rude. Some of the things touched on in this thread would mean more code paths, conditions, or other complexity to a pretty straightforward library.

**Comment by monolithed** [comment]
> I think I'll add make the options accessible through the helper's instance

Thanks

> @monolithed I don't really see a good use case for exposing things like extra , modifiers and elements right now. What did you have in mind?

It was just an example )

**Comment by AlexanderOtavka** [comment]
@marcohamersma I exposed `block.element()` and `block.modifiers()` because I thought it was needed for `ReactCSSTransitionGroup`.  However, it looks like that is actually unnecessary.  I'll take it out of #7.

@monolithed I'm a little skeptical of your general approach.  What if you need the transition name of your element to be in the form `"block__element"`.  Would you do `block.name + "__element"`?  It just seems to be a waste of the library.  Why not do:

```jsx
import * as ReactBlock from 'react-bem-helper';
import * as ReactCSSTransitionGroup from 'react-addons-css-transition-group';

class Preloader extends React.Component<void, void> {
	block = new ReactBlock('foo')

	render (): JSX.Element {
		return <div { ...this.block('', 'bar') } >
			<ReactCSSTransitionGroup transitionName={ this.block().className } >
				Loading...
			</ReactCSSTransitionGroup>
		</div>;
	}
}
```

It's only slightly longer than `this.block.name`, and allows you to do `this.block('element').className` in the future.

We can still expose the configuration of the block, but I would definitely suggest this approach for your use case.
